PR: https://github.com/odoo/odoo/pull/217987

From: 613e170853f606397dd1b238098628ed1998ebca
From: Stéphane Diez
Date: 2025-09-04 15:16:28

Breaking data model changes scores: del:2 + add:0, change matches:
-    lead_days_date = fields.Date(compute='_compute_lead_days')
-    visibility_days = fields.Float(

Total Changes: 85

[REF] (purchase,sale)_stock, mrp: use horizon_days

Remove visibility_days and replace it by horizon_days.
Also removing manufacturing_visibility_days and purchase_visibility_days.
horizon_days is a company setting where visibility_days was a system setting.
The default value for horizon_days is 365.
lead_days_date absorbs visibility_days_date and is now lead_horizon_date.

task 4779169

Part-of: odoo/odoo#217987
Related: odoo/enterprise#90534
Related: odoo/upgrade#8192
Signed-off-by: Tiffany Chang (tic) <tic@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock/models/res_company.py
+++ b/addons/stock/models/res_company.py
@@ -41,6 +41,10 @@ class ResCompany(models.Model):
         string='Day of the month', default=31,
         help="""Day of the month when the annual inventory should occur. If zero or negative, then the first day of the month will be selected instead.
         If greater than the last day of a month, then the last day of the month will be selected instead.""")
+    horizon_days = fields.Float(string="Replenishment Horizon", required=True, default=365,
+                                help="""Configure your horizon to trigger reordering rules earlier to get
+                                a head start on replenishment and avoid delays, or trigger it just-in-time
+                                ('0 days') to avoid overstocking.""")
 
     # used for sending stock text confirmation
     stock_text_confirmation = fields.Boolean("Stock Text Confirmation")

--- a/addons/stock/models/res_config_settings.py
+++ b/addons/stock/models/res_config_settings.py
@@ -56,6 +56,7 @@ class ResConfigSettings(models.TransientModel):
     replenish_on_order = fields.Boolean("Replenish on Order (MTO)", compute='_compute_replenish_on_order', inverse='_inverse_replenish_on_order')
     stock_text_confirmation = fields.Boolean(related='company_id.stock_text_confirmation', string='Stock Text Validation with stock move', readonly=False)
     stock_confirmation_type = fields.Selection(related='company_id.stock_confirmation_type', string='Stock Text Validation type', readonly=False)
+    horizon_days = fields.Float(related='company_id.horizon_days', readonly=False)
 
     def _compute_replenish_on_order(self):
         route = self.env.ref('stock.route_warehouse0_mto', raise_if_not_found=False)

--- a/addons/stock/models/stock_orderpoint.py
+++ b/addons/stock/models/stock_orderpoint.py
@@ -73,7 +73,7 @@ class StockWarehouseOrderpoint(models.Model):
     allowed_location_ids = fields.One2many(comodel_name='stock.location', compute='_compute_allowed_location_ids')
 
     rule_ids = fields.Many2many('stock.rule', string='Rules used', compute='_compute_rules')
-    lead_days_date = fields.Date(compute='_compute_lead_days')
+    lead_horizon_date = fields.Date(compute='_compute_lead_days')
     route_id = fields.Many2one(
         'stock.route', string='Route', domain="[('product_selectable', '=', True)]")
     qty_on_hand = fields.Float('On Hand', readonly=True, compute='_compute_qty', digits='Product Unit')
@@ -83,10 +83,6 @@ class StockWarehouseOrderpoint(models.Model):
     qty_to_order_manual = fields.Float('To Order Manual', digits='Product Unit')
 
     days_to_order = fields.Float(compute='_compute_days_to_order', help="Numbers of days  in advance that replenishments demands are created.")
-    visibility_days = fields.Float(
-        compute='_compute_visibility_days', inverse='_set_visibility_days', readonly=False,
-        help="Consider product forecast these many days in the future upon product replenishment, set to 0 for just-in-time.\n"
-             "The value depends on the type of the route (Buy or Manufacture)")
 
     unwanted_replenish = fields.Boolean('Unwanted Replenish', compute="_compute_unwanted_replenish")
     show_supply_warning = fields.Boolean(compute="_compute_show_supply_warning")
@@ -113,15 +109,14 @@ class StockWarehouseOrderpoint(models.Model):
         for orderpoint in self:
             orderpoint.show_supply_warning = not orderpoint.rule_ids
 
-    @api.depends('rule_ids', 'product_id.seller_ids', 'product_id.seller_ids.delay')
+    @api.depends('rule_ids', 'product_id.seller_ids', 'product_id.seller_ids.delay', 'company_id.horizon_days')
     def _compute_lead_days(self):
         orderpoints_to_compute = self.filtered(lambda orderpoint: orderpoint.product_id and orderpoint.location_id)
         for orderpoint in orderpoints_to_compute.with_context(bypass_delay_description=True):
             values = orderpoint._get_lead_days_values()
             lead_days, dummy = orderpoint.rule_ids._get_lead_days(orderpoint.product_id, **values)
-            lead_days_date = fields.Date.today() + relativedelta.relativedelta(days=lead_days['total_delay'])
-            orderpoint.lead_days_date = lead_days_date
-        (self - orderpoints_to_compute).lead_days_date = False
+            orderpoint.lead_horizon_date = fields.Date.today() + relativedelta.relativedelta(days=lead_days['total_delay'])
+        (self - orderpoints_to_compute).lead_horizon_date = False
 
     @api.depends('route_id', 'product_id', 'location_id', 'company_id', 'warehouse_id', 'product_id.route_ids')
     def _compute_rules(self):
@@ -151,13 +146,6 @@ class StockWarehouseOrderpoint(models.Model):
             if 'buy' in orderpoint.rule_ids.mapped('action'):
                 orderpoint.allowed_replenishment_uom_ids += orderpoint.product_id.seller_ids.product_uom_id
 
-    @api.depends('route_id', 'product_id')
-    def _compute_visibility_days(self):
-        self.visibility_days = 0
-
-    def _set_visibility_days(self):
-        return True
-
     @api.depends('route_id', 'product_id')
     def _compute_days_to_order(self):
         self.days_to_order = 0
@@ -226,10 +214,8 @@ class StockWarehouseOrderpoint(models.Model):
         action['context'] = {
             'active_id': self.product_id.id,
             'active_model': 'product.product',
-            'lead_days_date': format_date(self.env, self.lead_days_date),
-            'qty_to_order': self._get_qty_to_order(force_visibility_days=0),
-            'visibility_days_date': format_date(self.env, fields.Date.add(self.lead_days_date, days=int(self.visibility_days))),
-            'qty_to_order_with_visibility_days': self.qty_to_order_computed,
+            'lead_horizon_date': format_date(self.env, self.lead_horizon_date),
+            'qty_to_order': self._get_qty_to_order(),
         }
         warehouse = self.warehouse_id
         if warehouse:
@@ -328,12 +314,12 @@ class StockWarehouseOrderpoint(models.Model):
                 ]
 
     @api.depends('replenishment_uom_id', 'product_min_qty', 'product_max_qty',
-    'visibility_days', 'product_id', 'location_id', 'product_id.seller_ids.delay')
+    'product_id', 'location_id', 'product_id.seller_ids.delay', 'company_id.horizon_days')
     def _compute_qty_to_order_computed(self):
         def to_compute(orderpoint):
             rounding = orderpoint.product_uom.rounding
-            # The check is on purpose. We only want to consider the visibility days if the forecast is negative and
-            # there is a already something to ressuply base on lead times.
+            # The check is on purpose. We only want to consider the horizon days if the forecast is negative and
+            # there is already something to resupply base on lead times.
             return (
                 orderpoint.id
                 and float_compare(orderpoint.qty_forecast, orderpoint.product_min_qty, precision_rounding=rounding) < 0
@@ -352,22 +338,18 @@ class StockWarehouseOrderpoint(models.Model):
         """
         return False
 
-    def _get_qty_to_order(self, force_visibility_days=False, qty_in_progress_by_orderpoint=None):
+    def _get_qty_to_order(self, qty_in_progress_by_orderpoint=None):
         self.ensure_one()
-        visibility_days = self.visibility_days
-        if force_visibility_days is not False:
-            # Accepts falsy values such as 0.
-            visibility_days = force_visibility_days
         qty_to_order = 0.0
         qty_in_progress_by_orderpoint = qty_in_progress_by_orderpoint or {}
         qty_in_progress = qty_in_progress_by_orderpoint.get(self.id)
         if qty_in_progress is None:
             qty_in_progress = self._quantity_in_progress()[self.id]
         rounding = self.product_uom.rounding
-        # The check is on purpose. We only want to consider the visibility days if the forecast is negative and
-        # there is a already something to ressuply base on lead times.
+        # The check is on purpose. We only want to consider the horizon days if the forecast is negative and
+        # there is already something to resupply base on lead times.
         if float_compare(self.qty_forecast, self.product_min_qty, precision_rounding=rounding) < 0:
-            product_context = self._get_product_context(visibility_days=visibility_days)
+            product_context = self._get_product_context()
             qty_forecast_with_visibility = self.product_id.with_context(product_context).read(['virtual_available'])[0]['virtual_available'] + qty_in_progress
             qty_to_order = max(self.product_min_qty, self.product_max_qty) - qty_forecast_with_visibility
             qty_to_order = self._get_multiple_rounded_qty(qty_to_order)
@@ -395,12 +377,12 @@ class StockWarehouseOrderpoint(models.Model):
             'days_to_order': self.days_to_order,
         }
 
-    def _get_product_context(self, visibility_days=0):
+    def _get_product_context(self):
         """Used to call `virtual_available` when running an orderpoint."""
         self.ensure_one()
         return {
             'location': self.location_id.id,
-            'to_date': datetime.combine(self.lead_days_date + relativedelta.relativedelta(days=visibility_days), time.max)
+            'to_date': datetime.combine(self.lead_horizon_date, time.max)
         }
 
     def _get_orderpoint_action(self):
@@ -426,6 +408,8 @@ class StockWarehouseOrderpoint(models.Model):
         # Remove previous automatically created orderpoint that has been refilled.
         orderpoints_removed = orderpoints._unlink_processed_orderpoints()
         orderpoints = orderpoints - orderpoints_removed
+        if self.env.context.get('force_orderpoint_recompute', False):
+            orderpoints._compute_qty_to_order_computed()
         to_refill = defaultdict(float)
         all_product_ids = self._get_orderpoint_products()
         all_replenish_location_ids = self._get_orderpoint_locations()
@@ -640,9 +624,9 @@ class StockWarehouseOrderpoint(models.Model):
                             origin = orderpoint.name
                         if orderpoint.product_uom.compare(orderpoint.qty_to_order, 0.0) == 1:
                             date = orderpoint._get_orderpoint_procurement_date()
-                            global_visibility_days = self.env.context.get('global_visibility_days', self.env['ir.config_parameter'].sudo().get_param('stock.visibility_days', 0))
-                            if global_visibility_days:
-                                date -= relativedelta.relativedelta(days=int(global_visibility_days))
+                            global_horizon_days = orderpoint.get_horizon_days()
+                            if global_horizon_days:
+                                date -= relativedelta.relativedelta(days=int(global_horizon_days))
                             values = orderpoint._prepare_procurement_values(date=date)
                             procurements.append(self.env['procurement.group'].Procurement(
                                 orderpoint.product_id, orderpoint.qty_to_order, orderpoint.product_uom,
@@ -700,7 +684,7 @@ class StockWarehouseOrderpoint(models.Model):
         return True
 
     def _get_orderpoint_procurement_date(self):
-        return timezone(self.company_id.partner_id.tz or 'UTC').localize(datetime.combine(self.lead_days_date, time(12))).astimezone(UTC).replace(tzinfo=None)
+        return timezone(self.company_id.partner_id.tz or 'UTC').localize(datetime.combine(self.lead_horizon_date, time(12))).astimezone(UTC).replace(tzinfo=None)
 
     def _get_orderpoint_products(self):
         return self.env['product.product'].search([('is_storable', '=', True), ('stock_move_ids', '!=', False)])
@@ -717,6 +701,11 @@ class StockWarehouseOrderpoint(models.Model):
             qty_to_order = replenishment_multiple._compute_quantity(qty_to_order, self.product_id.uom_id)
         return qty_to_order
 
-    @api.model
-    def get_visibility_days(self):
-        return self.env['ir.config_parameter'].sudo().get_param('stock.visibility_days', 0)
+    def get_horizon_days(self):
+        """ Return the value for Horizon. This can be (in order of priority):
+        - the value set in context in the replenishment view
+        - the value set on the company of the all the records in self. In multi-company, it will take the biggest value.
+        - the value set on the company of the user if all else fail.
+        """
+        max_company_horizon_days = max(self.company_id.mapped('horizon_days')) if self.company_id else 0
+        return self.env.context.get('global_horizon_days', max_company_horizon_days or self.env.company.horizon_days)

--- a/addons/stock/models/stock_rule.py
+++ b/addons/stock/models/stock_rule.py
@@ -415,9 +415,10 @@ class StockRule(models.Model):
         delays = defaultdict(float)
         delay = sum(self.filtered(lambda r: r.action in ['pull', 'pull_push']).mapped('delay'))
         delays['total_delay'] += delay
-        global_visibility_days = self.env.context.get('global_visibility_days', self.env['ir.config_parameter'].sudo().get_param('stock.visibility_days', 0))
-        if global_visibility_days:
-            delays['total_delay'] += int(global_visibility_days)
+        max_company_horizon_days = max(self.company_id.mapped('horizon_days')) if self.company_id else 0
+        global_horizon_days = self.env.context.get('global_horizon_days', max_company_horizon_days or self.env.company.horizon_days)
+        if global_horizon_days:
+            delays['total_delay'] += int(global_horizon_days)
         if self.env.context.get('bypass_delay_description'):
             delay_description = []
         else:
@@ -426,8 +427,8 @@ class StockRule(models.Model):
                 for rule in self
                 if rule.action in ['pull', 'pull_push'] and rule.delay
             ]
-        if global_visibility_days:
-            delay_description.append((_('Time Horizon'), _('+ %d day(s)', int(global_visibility_days))))
+        if global_horizon_days:
+            delay_description.append((_('Time Horizon'), _('+ %d day(s)', int(global_horizon_days))))
         return delays, delay_description
 
 
